modules/trainer/fs_incremental_trainer.py

```diff
@@ -83,7 +80,112 @@ def __init__(self, cfg, backbone_net, post_processor, criterion, dataset_module,
+                k_far = {}
+            if 'close' in baseset_type:
+                m_close = m
+                similarity_closest_dic = {}
+                k_close = {}
+            if baseset_type == 'far_close':
+                m_close //= 2
+                m_far //= 2
+            mean_weight_dic = {}
+
+            # Get the mean weight of each class
+            for c in base_id_list:
+                mean_weight_dic[c] = self.prv_post_processor.pixel_classifier.class_mat.weight.data[c]
+                mean_weight_dic[c] = mean_weight_dic[c].reshape((-1))
+                mean_weight_dic[c] = mean_weight_dic[c].cpu().unsqueeze(0)
+
+            for c in base_id_set:
+                if 'far' in baseset_type:
+                    k_far[c] = min(m_far, len(self.train_set_vanilla_label.dataset.get_class_map(c)))
+                    similarity_farthest_dic[c] = []
+                if 'close' in baseset_type:
+                    k_close[c] = min(m_close, len(self.train_set_vanilla_label.dataset.get_class_map(c)))
+                    similarity_closest_dic[c] = []
+                    
+            # Maintain a m-size heap to store the top m images of each class
+            for i in tqdm(range(len(self.train_set_vanilla_label))):
+                img, mask = self.train_set_vanilla_label[i]
+                class_list = torch.unique(mask).tolist()
+                img_tensor = torch.stack([img]).cuda()
+                mask_tensor = torch.stack([mask]).cuda()
+                for c in class_list:
+                    if c not in base_id_set:
+                        continue
+                    with torch.no_grad():
+                        img_feature = self.prv_backbone_net(img_tensor)
+                        img_weight = utils.masked_average_pooling(mask_tensor == c, img_feature, True)
+                    img_weight = img_weight.cpu().unsqueeze(0)
+                    similarity = F.cosine_similarity(img_weight, mean_weight_dic[c])
+                    # Update closest heap
+                    if 'close' in baseset_type:
+                        if len(similarity_closest_dic[c]) < k_close[c]:
+                            heapq.heappush(similarity_closest_dic[c], ((similarity, i)))
+                        else:
+                            heapq.heappushpop(similarity_closest_dic[c], ((similarity, i)))
+                    # Update farthest heap
+                    if 'far' in baseset_type:
+                        if len(similarity_farthest_dic[c]) < k_far[c]:
+                            heapq.heappush(similarity_farthest_dic[c], ((-similarity, i)))
+                        else:
+                            heapq.heappushpop(similarity_farthest_dic[c], ((-similarity, i)))   
+            if not os.path.exists(baseset_folder):   
+                os.makedirs(baseset_folder)  
+            if 'close' in baseset_type:
+                torch.save(similarity_closest_dic, f"{baseset_folder}/similarity_closest_dic_{baseset_type}")
```

Comment:
This needs to consider splits. Otherwise there may be information leakage. A suggestion is to use self.cfg.task_name as prefix when saving/loading the pre-computed embeddings.